internal/plugin/invoke.go

```diff
@@ -0,0 +1,203 @@
+	Type component.Type
+	Name string
+
+	Dir string
+
+	ConfigData []byte
+	JsonConfig bool
+}
+
+func startInstance(
+	ctx context.Context, log hclog.Logger, req *PluginRequest,
+) (*Instance, error) {
+	// Figure out the command to execute to start to bring the plugin
+	// into existance
+	pluginPaths, err := DefaultPaths(req.Dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var fn interface{}
+
+	cmd, err := Discover(&req.Config, pluginPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmd != nil {
+		fn = Factory(cmd, req.Type)
+	} else {
+		// If the plugin was not found, it is only an error if
+		// we don't have that plugin as a builtin plugin (for builtin plugins
+		// we relaunch the existing process)
+
+		if _, ok := Builtins[req.Config.Name]; !ok {
+			return nil, fmt.Errorf("plugin %q not found", req.Config.Name)
+		} else {
+			log.Debug("plugin found as builtin")
+			fn = BuiltinFactory(req.Config.Name, req.Type)
+		}
+	}
+
+	// Use argmapper to call our plugin creation function with
+	// the standard context and log args. This is just to create
+	// the plugin process via go-plugin, not to actually invoke
+	// anything within the plugin yet.
+
+	afn, err := argmapper.NewFunc(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	fnResult := afn.Call(argmapper.Typed(ctx, log))
+	if err := fnResult.Err(); err != nil {
+		return nil, err
+	}
+
+	rawComponent := fnResult.Out(0)
+
+	// If we have a plugin.Instance then we can extract other information
+	// from this plugin. We accept pure factories too that don't return
+	// this so we type-check here. This would only happen if the above code
+	// used a creation function other than the *Factory functions.
+	pi, ok := rawComponent.(*Instance)
+	if !ok {
+		pi = &Instance{
+			Component: rawComponent,
+			Close:     func() {},
+		}
+	}
+
+	return pi, nil
+}
+
+// Plugin is the state of a created plugin to be invoked, returned by OpenPlugin.
+type Plugin struct {
+	req      *PluginRequest
+	Instance *Instance
+}
+
+// OpenPlugin resolves the plugin information in req and returns a Plugin value
+// to have Invoke called upon it. OpenPlugin also returns the raw component
+// interface, which is used to get the function value that will be invoked
+func OpenPlugin(
+	ctx context.Context, log hclog.Logger, req *PluginRequest,
+) (*Plugin, interface{}, error) {
+	pi, err := startInstance(ctx, log, req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var hclCtx *hcl.EvalContext
+
+	log.Debug("decoding task plugin config")
+
+	var (
+		file  *hcl.File
+		diags hcl.Diagnostics
+	)
+
+	if req.JsonConfig {
+		file, diags = json.Parse(req.ConfigData, "plugin.json")
+	} else {
+		file, diags = hclsyntax.ParseConfig(req.ConfigData, "plugin.hcl", hcl.Pos{Line: 1, Column: 1})
+	}
+
+	if diags.HasErrors() {
+		return nil, nil, diags
+	}
+
+	if file.Body != nil {
+		component.Configure(pi.Component, file.Body, hclCtx.NewChild())
+	}
+
+	return &Plugin{req: req, Instance: pi}, pi.Component, nil
```

Comment:
Feel like we should just return the `*Plugin, error` tuple and maybe introduce a helper `func (p *Plugin) Component()` to access `pi.Component` since that's common. That feels like it's more Go-ish.

Reply:
Yeah, I had that originally, but you have to dig it out as `p.Instance.Component` each time which during experimentation was really long given you HAVE to use that value to actually use the plugin interface. I also experiment with having `p.C()` ie, a C() function on Plugin to return the component. It felt cleaner to just return the value from Open.

Reply:
Okay that's fine